R/condition_pr_matrix.R

```diff
@@ -0,0 +1,197 @@
+#' Build condition probaability matrix for Horvitz-Thompson estimation from randomization declaration
+#'
+#' @param declaration An object of class 'ra_declaration' that contains the experimental design
+#'
+#' @export
+declaration_to_condition_pr_mat <- function(declaration) {
+
+  if (class(declaration) != 'ra_declaration') {
+    stop("declaration must be an object of class 'ra_declaration'")
+  }
+
+  if (declaration$cleaned_arguments$num_arms > 2) {
+    stop("The 'declaration' argument can only be used with a binary treatment variable.")
+  }
+
+  declaration_call <- as.list(declaration$original_call)
+  simple <- eval(declaration_call$simple)
+
+  if (declaration$ra_type == 'simple') {
+
+    mat_00 <- tcrossprod(declaration$probabilities_matrix[, 1])
+    diag(mat_00) <- declaration$probabilities_matrix[, 1]
+    mat_11 <- tcrossprod(declaration$probabilities_matrix[, 2])
+    diag(mat_11) <- declaration$probabilities_matrix[, 2]
+
+    joint_probs <- tcrossprod(declaration$probabilities_matrix[, 1],
+                              declaration$probabilities_matrix[, 2])
+
+    mat_01 <- joint_probs
+    mat_01 <- copy_upper_to_lower_triangle(mat_01)
+    mat_10 <- joint_probs
+    mat_10 <- copy_lower_to_upper_triangle(mat_10)
+    diag(mat_01) <- diag(mat_10) <- 0
+
+    condition_pr_matrix <-
+      rbind(cbind(mat_00, mat_01),
+            cbind(mat_10, mat_11))
```

Comment:
Thanks, over 2x faster. A lot of this code is written very transparently because I was still struggling to understand how HT works as I was working through. Definitely optimization to be done throughout this.